src/Backend/Modules/MediaLibrary/Ajax/UploadMediaItem.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+namespace Backend\Modules\MediaLibrary\Ajax;
+
+use Backend\Core\Engine\Authentication as BackendAuthentication;
+use Backend\Core\Engine\Base\AjaxAction as BackendBaseAJAXAction;
+use Backend\Core\Language\Language;
+use Backend\Modules\MediaLibrary\Domain\MediaItem\Command\CreateMediaItemFromSource;
+use Backend\Modules\MediaLibrary\Domain\MediaFolder\MediaFolder;
+use Backend\Modules\MediaLibrary\Domain\MediaItem\Event\MediaItemCreated;
+use Backend\Modules\MediaLibrary\Domain\MediaItem\MediaItem;
+
+/**
+ * This AJAX-action is being used to upload new MediaItem items and save them into to the database.
+ */
+class UploadMediaItem extends BackendBaseAJAXAction
+{
+    // override existing media
+    const OVERRIDE_EXISTING = false;
+
+    /**
+     * Execute the action
+     */
+    public function execute()
+    {
+        parent::execute();
+
+        try {
+            // 60 minutes execution time
+            set_time_limit(60 * 60);
+        } catch (\Exception $e) {
+            // Do nothing
+        }
+
+        // increase memory limit
+        ini_set('memory_limit', '128M');
+
+        $contentType = null;
+
+        // define folder id
+        $folderId = trim(\SpoonFilter::getPostValue(
+            'folder_id',
+            null,
+            '',
+            'int'
+        ));
+
+        // define destination URL
+        $destinationURL = trim(\SpoonFilter::getPostValue(
+            'filename',
+            null,
+            '',
+            'string'
+        ));
+
+        // redefine destination URL as unique URL
+        $destinationURL = $this->get('media_library.manager.file')->getUniqueURL(
+            $destinationURL,
+            self::OVERRIDE_EXISTING
+        );
+
+        // define destination source path
+        $destinationSourcePath = MediaItem::getUploadRootDir() . '/' . $destinationURL;
+
+        // create folder if not exists
+        if (!\SpoonDirectory::exists(dirname($destinationSourcePath))) {
+            \SpoonDirectory::create(dirname($destinationSourcePath));
+        }
+
+        // HTTP headers for no cache etc
+        header('Expires: Mon, 26 Jul 1997 05:00:00 GMT');
+        header('Last-Modified: ' . gmdate('D, d M Y H:i:s') . ' GMT');
+        header('Cache-Control: no-store, no-cache, must-revalidate');
+        header('Cache-Control: post-check=0, pre-check=0', false);
+        header('Pragma: no-cache');
+
+        // get parameters
+        $chunk = isset($_REQUEST['chunk']) ? (int) $_REQUEST['chunk'] : 0;
+        $chunks = isset($_REQUEST['chunks']) ? (int) $_REQUEST['chunks'] : 0;
+
+        // look for the content type header
+        if (isset($_SERVER['HTTP_CONTENT_TYPE'])) {
+            $contentType = $_SERVER['HTTP_CONTENT_TYPE'];
+        }
+        if (isset($_SERVER['CONTENT_TYPE'])) {
+            $contentType = $_SERVER['CONTENT_TYPE'];
+        }
+
+        // handle multipart files
+        if (strpos($contentType, 'multipart') !== false) {
```

Comment:
I'm rather sure that this can be refactored into cleaner more easy to understand code, don't be afraid to use a lot of small methods